Refactor App & add some documentation

DarthSim opened the pull request:
What I've done here:

- Extracted messaging logic into `SlackClient`.
- Extracted webhook logic into `CallbbackServer`.
- Extracted request management & caching logic into `RequestManager`.
- Added documentation for the most functions and types.

Extracting things from `App` reduced its responsibility and reduced the main file size almost by half.

What do you think?

marshall-lee commented:
Good job! But it'd be better to all the things incrementally.

- Start with the extraction of http server. This is most straightforward since it does not depend on api client and cache.
- Don't bother with with `request_manager` extraction since it's another wrapper for access client. There'll be no need in it whenever https://github.com/gravitational/teleport/issues/3286 is resolved
- `slack_client` is worth extracting too but lets do it last. It's not very clear for me now how it should look with the absence of cache `Entry`.

DarthSim commented on access/slackbot/slackbot.go:
```diff
@@ -159,440 +147,187 @@ func NewApp(conf Config) (*App, error) {
-}
+		watcherErr = trace.Wrap(
+			// Start watching for request events with a.OnPendingRequest and a.OnDeletedRequest as callback functions
+			a.requestManager.Watch(ctx, a.OnPendingRequest, a.OnDeletedRequest),
```
What do you mean?

marshall-lee replied:
ah i just forgot to delete this comment

DarthSim commented on access/slackbot/cache.go:
```diff
@@ -36,15 +36,15 @@ type Entry struct {
 type RequestCache struct {
 	sync.Mutex
-	entries map[string]Entry
+	entries map[string]*Entry
```
1. Since we only put and pop entries, there is no danger in sharing the pointer. Many caching solutions like `sync.Pool` has a recommendation to use pointers.
2. It causes lots of redundant allocations.
3. Well, yes, but here we allocate memory even when we missed the cache, which is no good.

> No need to touch this file anyway, eventually it will be completely removed when gravitational/teleport#3286 is done.

Can't disagree here :)

marshall-lee replied:
> It causes lots of redundant allocations.

Stack allocations are a lot cheapier than heap allocations. And it's also expected to be safe as long as resulting structure is not very big. For example, look at what code is being generated here https://stackoverflow.com/questions/10866195/stack-vs-heap-allocation-of-structs-in-go-and-how-they-relate-to-garbage-collec

DarthSim replied:
I made some benchmarks and yep, it looks like there are no heap allocations, though I was pretty sure they should be.
Solution with pointers is a bit faster but it's not critical here.

DarthSim commented on access/slackbot/callback_server.go:
```diff
@@ -0,0 +1,118 @@
+// CallbackServer is a wrapper around http.Server that processes Slack interaction events.
+// It verifies incoming requests and calls onCallback for valid ones
+type CallbackServer struct {
+	ctx        context.Context
```
I made this so `CallbackServer` would match `http.Handler` interface. This is allowed by "Code Review Comments":

> The one exception is for methods whose signature must match an interface in the standard library or in a third party library.

But I forgot about `http.HandlerFunc` interface which is a better use here. Fixed.

DarthSim commented:
> `slack_client` is worth extracting too but lets do it last. It's not very clear for me now how it should look with the absence of cache `Entry`.

We'll just replace `Entry` with `access.Request` since it will contain all necessary data.

fspmarshall commented on access/slackbot/callback_server.go:
```diff
@@ -0,0 +1,118 @@
+	if len(cb.ActionCallback.BlockActions) != 1 {
+		http.Error(rw, "expected exactly one block action", 500)
+		return
+	}
```
Likely symptomatic of a bug, lets log this.

DarthSim replied:
Done

DarthSim commented:
I removed `RequestManager` and made some changes that will help us to migrate to ext-data smoothly:

- I rewrote `SlackClient` a bit so it works with `access.Request`.
- I removed the cache usage from `loadRequest`. We don't actually need it there, and we'd drop its usage after migrating to ext-data anyway. Instead, I drop the cache record explicitly in `OnSlackAction`.

The only place where currently we need cache (or ext-data in future) is `OnDeletedRequest`, and I'm a bit puzzled about its future. `OpDeleted` event gives us only request ID, so we don't know if this request has been already processed. If `OpDeleted` won't return ext-data, we'll have quite a problem here.
